Implement private and shareable ipc modes

kolyshkin opened the pull request:
Since the commit d88fe447df0e8 ("Add support for sharing /dev/shm/ and
/dev/mqueue between containers") container's /dev/shm is mounted on the
host first, then bind-mounted inside the container. This is done that
way in order to be able to share this container's IPC namespace
(and the /dev/shm mount point) with another container.

Unfortunately, this functionality breaks container checkpoint/restore
(even if IPC is not shared). Since /dev/shm is an external mount, its
contents is not saved by `criu checkpoint`, and so upon restore any
application that tries to access data under /dev/shm is severily
disappointed as the data is not there (which usually results in a fatal crash). For example, see
https://github.com/xemul/criu/issues/300 and https://github.com/xemul/criu/issues/320.

This commit solves the issue by introducing new IPC modes for containers
(in addition to `host` and `container:ID`). The new modes are:

 - `shareable`:	enables sharing this container's IPC with others
		(this used to be a default behavior)

 - `private`:	disables sharing this container's IPC.

In `private` mode, container's /dev/shm is truly mounted inside the
container, without any bind-mounting from the host, which solves the
issue for checkpoint/restore.

Surely, to ultimately solve it, we'd need to make `private` the default
mode, but it breaks the backward compatibility. So, let's make the
default container IPC mode per-daemon configurable (with built-in
default set to `shareable` for now). The default can be changed
either via a daemon CLI option (`--default-shm-mode`) or configuration
file parameter of the same name.

Surely, one can _only_ set either `shareable` or `private` IPC modes as a
daemon default. Other modes (like 'host' or 'container') do not make
sense in this context and are therefore not allowed.

Some other changes this patch introduces are:

1. A mount for /dev/shm is added to default OCI Linux spec.

2. `IpcMode.Valid()` is simplified to remove duplicated code that parsed
   'container:ID' form. Note the old version used to check that ID does
   not contain a semicolon -- this is no longer the case (tests are
   modified accordingly). The motivation is we should either do a
   proper check for container ID validity, or don't check it at all
   (since it is checked in other places anyway). I chose the latter.

3. IPC mode tests are modified to add checks for newly added values.

I did some manual testing, and it appears to be working. 

~~TODO: write integration tests~~
Test cases come in the second patch, and they cover the scenarios I can think of.

kolyshkin committed:
Implement none, private, and shareable ipc modes

Since the commit d88fe447df0e8 ("Add support for sharing /dev/shm/ and
/dev/mqueue between containers") container's /dev/shm is mounted on the
host first, then bind-mounted inside the container. This is done that
way in order to be able to share this container's IPC namespace
(and the /dev/shm mount point) with another container.

Unfortunately, this functionality breaks container checkpoint/restore
(even if IPC is not shared). Since /dev/shm is an external mount, its
contents is not saved by `criu checkpoint`, and so upon restore any
application that tries to access data under /dev/shm is severily
disappointed (which usually results in a fatal crash).

This commit solves the issue by introducing new IPC modes for containers
(in addition to 'host' and 'container:ID'). The new modes are:

 - 'shareable':	enables sharing this container's IPC with others
		(this used to be the implicit default);

 - 'private':	disables sharing this container's IPC.

In 'private' mode, container's /dev/shm is truly mounted inside the
container, without any bind-mounting from the host, which solves the
issue.

While at it, let's also implement 'none' mode. The motivation, as
eloquently put by Justin Cormack, is:

> I wondered a while back about having a none shm mode, as currently it is
> not possible to have a totally unwriteable container as there is always
> a /dev/shm writeable mount. It is a bit of a niche case (and clearly
> should never be allowed to be daemon default) but it would be trivial to
> add now so maybe we should...

...so here's yet yet another mode:

 - 'none':	no /dev/shm mount inside the container (though it still
		has its own private IPC namespace).

Now, to ultimately solve the abovementioned checkpoint/restore issue, we'd
need to make 'private' the default mode, but unfortunately it breaks the
backward compatibility. So, let's make the default container IPC mode
per-daemon configurable (with the built-in default set to 'shareable'
for now). The default can be changed either via a daemon CLI option
(--default-shm-mode) or a daemon.json configuration file parameter
of the same name.

Note one can only set either 'shareable' or 'private' IPC modes as a
daemon default (i.e. in this context 'host', 'container', or 'none'
do not make much sense).

Some other changes this patch introduces are:

1. A mount for /dev/shm is added to default OCI Linux spec.

2. IpcMode.Valid() is simplified to remove duplicated code that parsed
   'container:ID' form. Note the old version used to check that ID does
   not contain a semicolon -- this is no longer the case (tests are
   modified accordingly). The motivation is we should either do a
   proper check for container ID validity, or don't check it at all
   (since it is checked in other places anyway). I chose the latter.

3. IpcMode.Container() is modified to not return container ID if the
   mode value does not start with "container:", unifying the check to
   be the same as in IpcMode.IsContainer().

3. IPC mode unit tests (runconfig/hostconfig_test.go) are modified
   to add checks for newly added values.

[v2: addressed review at https://github.com/moby/moby/pull/34087#pullrequestreview-51345997]
[v3: addressed review at https://github.com/moby/moby/pull/34087#pullrequestreview-53902833]
[v4: addressed the case of upgrading from older daemon, in this case
     container.HostConfig.IpcMode is unset and this is valid]
[v5: document old and new IpcMode values in api/swagger.yaml]
[v6: add the 'none' mode, changelog entry to docs/api/version-history.md]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
From a quick look, Windows test failures does not appear to be related to my changes

kolyshkin committed:
Test cases for new ipc modes

These test cases cover various arguments for docker create/run --ipc
option, as well as daemon's --default-ipc-mode cli option and
configuration file parameter.

For the description of container IPC modes, see previous commit.

To run these:

	TESTFLAGS='-check.f IpcMode' make test-integration-cli

[v2: simplify TestDaemonEvents(), add default-ipc-mode presense check]
[v3: add TestDaemonIpcModeVSRestart]
[v4: ipcmode test now uses client lib instead of CLI (except for exec)]
[v5: nitpicks in comments]
[v6: add test case for "none"; fix a typo; simplify TestDaemonEvents() more]

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
ping @thaJeztah @cpuguy83 @justincormack @rhatdan @mrunalp

thaJeztah commented on runconfig/hostconfig_test.go:
```diff
@@ -60,16 +60,30 @@ func TestNetworkModeTest(t *testing.T) {
-		"container:name:something": {false, false, true, false},
-		"container:":               {false, false, true, false},
+		// private, host, container, shareable, valid
+		"":                {false, false, false, false, true},
```
nit: would it be more readable to use a struct for these? The second loop for parsing container names could be combined in that case;

```go
func TestIpcModeTest(t *testing.T) {
	ipcModes := map[container.IpcMode]struct {
		private   bool
		host      bool
		container bool
		shareable bool
		valid     bool
		ctrName   string
	}{
		"":                      {valid: true},
		"private":               {private: true, valid: true},
		"something":             {},
		"something:weird":       {ctrName: "weird"},
		":weird":                {ctrName: "weird"},
		"host":                  {host: true, valid: true},
		"container":             {},
		"container:name":        {container: true, ctrName: "name", valid: true},
		"container:name1:name2": {container: true, ctrName: "name1:name2", valid: true},
		"container:":            {container: true, valid: true},
		"shareable":             {shareable: true, valid: true},
	}
	for ipcMode, state := range ipcModes {
		assert.Equal(t, state.private, ipcMode.IsPrivate(), "IpcMode.IsPrivate() parsing failed for %q", ipcMode)
		assert.Equal(t, state.host, ipcMode.IsHost(), "IpcMode.IsHost()  parsing failed for %q", ipcMode)
		assert.Equal(t, state.container, ipcMode.IsContainer(), "IpcMode.IsContainer()  parsing failed for %q", ipcMode)
		assert.Equal(t, state.shareable, ipcMode.IsShareable(), "IpcMode.IsShareable()  parsing failed for %q", ipcMode)
		assert.Equal(t, state.valid, ipcMode.Valid(), "IpcMode.Valid()  parsing failed for %q", ipcMode)
		assert.Equal(t, state.ctrName, ipcMode.Container(), "IpcMode.Container() parsing failed for %q", ipcMode)
	}
}
```

thaJeztah reviewed (changes requested):
Thanks @kolyshkin - I left some comments (but interested in other reviews as well)

Wondering would this also reduce the chance of `/shm` "device in use" problems?

rhatdan commented:
The way we have gone after this is to add a RUNC hook oci-umount which cleans up leaked mount points into a container.

https://github.com/projectatomic/oci-umount

kolyshkin committed:
Use container.HostConfig.ShmSize directly

It was noted[1] that container's HostConfig.ShmSize, if not set, should be
initialized to daemon default value during container creation.

In fact, it is already done in daemon.adaptContainerSettings, so we can use
value from container.HostConfig directly.

[1] https://github.com/moby/moby/pull/34087#discussion_r128656429

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@thaJeztah thank you for review! I think I addressed all your comments in the updated patches, can you please take another look?

thaJeztah commented:
Thank you @kolyshkin ! I gave it a quick glance, and definitely looks better; I'll do a full review soon.

kolyshkin commented:
@thaJeztah can you please take another look?

thaJeztah commented:
oops, it slipped under my radar; will have a look 👍

thaJeztah commented on integration-cli/docker_api_ipcmode_test.go:
```diff
@@ -0,0 +1,207 @@
+	testIpcPrivateShareable(c, "shareable", true)
+}
+
+// this is used to test --ipc container:NNN mode in various scenarios
```
perhaps `s/this/testIpcContainer/`, e.g.

```
// testIpcContainer is a helper function to test --ipc container:NNN mode in various scenarios
```

kolyshkin replied:
I thought this particular way of writing comments is reserved for public functions, as those comments make its way to code documentation (via godoc). This is a private function, and the comment is for the sake of code reader, not the reader of what godoc has produced.

Anyway I'll fix it.

thaJeztah commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2985,6 +2985,165 @@ func (s *DockerDaemonSuite) TestShmSizeReload(c *check.C) {
 	c.Assert(strings.TrimSpace(out), check.Equals, fmt.Sprintf("%v", size))
 }
 
+// this is used to test both "private" and "shareable" daemon default ipc modes
```
```
testDaemonIpcPrivateShareable is a helper function to..
```

helper/utility whatever  😊

thaJeztah reviewed (changes requested):
Thanks @kolyshkin - left some comments, but could surely use some more eyes on this (for areas I'm not fully familiar with 😅 )

kolyshkin commented on daemon/config/config_unix.go:
```diff
@@ -61,3 +68,24 @@ func (conf *Config) IsSwarmCompatible() error {
+		return fmt.Errorf("Default IPC mode setting (%v) is invalid. "+hint, dm)
+	}
+	if dm != "" && !dm.IsPrivate() && !dm.IsShareable() {
+		return fmt.Errorf("IPC mode \"%v\" is not supported as default value. "+hint, dm)
```
I'd rather leave it as is, since the error diagnostics is cleaner if we distinguish between an invalid value and a value that is valid but can not be used as default. Surely, it's a minor thing so whatever you say.

kolyshkin commented on daemon/container_operations_unix.go:
```diff
@@ -63,7 +63,11 @@ func (daemon *Daemon) getIpcContainer(container *container.Container) (*containe
 	if err != nil {
```
We need access to supplied container's hostconfig (to call its IpcMode.Container() method), I can't see how this can be done with the change you propose.

kolyshkin commented on daemon/container_operations_unix.go:
```diff
@@ -63,7 +63,11 @@ func (daemon *Daemon) getIpcContainer(container *container.Container) (*containe
 	if err != nil {
 		return nil, errors.Wrapf(err, "cannot join IPC of a non running container: %s", container.ID)
```
> does daemon.GetContainer(containerID) actually produce an error if the container is not running?

No. This is checked later by daemon.CheckContainer()). I guess you were misled by the error message ("cannot join IPC of a non running container"), I am going to change it to:
```go
return nil, errors.Wrapf(err, "cannot join IPC of nonexistent container: %s", donorID)
```

> Testing on current docker, it looks like not;
> ...
> docker: Error response from daemon: linux spec namespaces: container 861827549edaa5c9fbd08f72bd621d1ae8f5c688d8e89d4c35d9437c4a160537 is not running.

This is the error message from the abovementioned daemon.CheckContainer().

> Paused containers are ok?
Yes.

> I also notice that the container argument is being re-used here
Oh what a mess, I should have noticed it too. Fixed.

kolyshkin commented on daemon/container_operations_unix.go:
```diff
@@ -63,7 +63,11 @@ func (daemon *Daemon) getIpcContainer(container *container.Container) (*containe
 	}
-	return container, daemon.checkContainer(container, containerIsRunning, containerIsNotRestarting)
+	// Check that "donor" container ipc is shareable
+	if st, e := os.Stat(container.ShmPath); e == nil && st.IsDir() {
```
os.Stat() can return so many errors, but in most cases it will be ENOENT, to which we should say "container IPC is not shareable". All other errors, well, fall under "unexpected error/this should not happen".

I changed the code based on that.

kolyshkin commented on daemon/container_operations_unix.go:
```diff
@@ -120,19 +122,21 @@ func (daemon *Daemon) setupIpcDirs(c *container.Container) error {
 				return err
 			}
 
-			shmSize := int64(daemon.configStore.ShmSize)
```
As the commit message says, this code is redundant (not needed). c.HostConfig.ShmSize has already been set to daemon default value, it happens in daemon.adaptContainerSettings, so here we can just use c.HostConfig.ShmSize directly.

thaJeztah replied:
Ah sorry; late night code reviews .. 😅

kolyshkin commented on daemon/container_operations_unix.go:
```diff
@@ -120,19 +122,21 @@ func (daemon *Daemon) setupIpcDirs(c *container.Container) error {
+	case ipcMode.IsPrivate():
+		c.ShmPath = ""
+
+	default:
```
In case of private ipc mode, this variable is not used (as it is only used to do bind-mounting). If it is used, that'd be an error. I explicitly set it to "" here for the sake of catching such potential misuse. Surely, this assignment is optional, but I'd rather leave it as.

thaJeztah replied:
Oh, apologies, I meant the `default:` case, because this PR invalidates what was previously valid, causing upgrades of docker to fail. Here's to illustrate what I meant:

Build from master, then:

```bash
$ docker run -dit --name foo busybox
$ docker inspect -f '{{json .HostConfig.IpcMode}}' foo
""

$ docker stop foo
```

Stop the daemon, build from your PR and start the new daemon

```bash
$ docker start foo
Error response from daemon: linux spec namespaces: Invalid IPC mode:
Error: failed to start containers: foo
```

The `foo` container was created with the "old" (master) version of Docker, but after upgrading to the new (this pr) version of docker, the container can no longer be started

kolyshkin replied:
Got it, thank you so much for pointing this out, I totally overlooked this scenario :-\

By the way, is it possible to change existing container's IpcMode (other than the direct editing of its hostconfig.json)? I am thinking about something like `docker set` which can be used to apply parameters to an existing container (surely, some can be applied runtime, some others like --ipc should require restart).

thaJeztah replied:
We do have `docker update`, but it's a curated list of options that we allow updating; generally, containers should be considered immutable, so re-creating the container is in most cases the answer. For `docker update` we (so far) limited options to updating resource-constraints, e.g. update memory-limits for an existing container (as that can be considered an operational concern)

kolyshkin replied:
I am asking because, even with the new daemon and the default-ipc-mode set to 'private' the original problem (of container's ability to correctly checkpoint/restore) won't be fixed because of such pre-existing containers for which we apply the old default.

I'm not quite sure how to address that. One way is to add `--ipc` to `docker update` (with a requirement for a container to be stopped). Another way is to enforce the current daemon default (instead of falling back to "shareable" the way it is now), but it might break backward compatibility.

Any thoughts?

kolyshkin commented on daemon/daemon_unix.go:
```diff
@@ -596,6 +609,10 @@ func (daemon *Daemon) reloadPlatform(conf *config.Config, attributes map[string]
 		daemon.configStore.ShmSize = conf.ShmSize
 	}
 
+	if conf.IpcMode != "" {
```
I fail to see the reason why a simple comparison is not enough, and what can be wrong if we use it instead of conf.IsValueSet(). From what I see, it makes sense to use it to distinguish the case when a parameter is not set from the case when a parameter is set to `nul`. Maybe I'm missing something?

thaJeztah replied:
Ah, yes, I was looking at the code above, but agree it doesn't add much

kolyshkin commented on container/container_unix.go:
```diff
@@ -171,26 +171,25 @@ func (container *Container) HasMountFor(path string) bool {
+	shmPath, err := container.ShmResourcePath()
+	if err != nil {
+		logrus.Error(err)
+		warnings = append(warnings, err.Error())
```
Yes, it is from the time when the code dealt with /dev/shm and /dev/mqueue, the latter was later dropped. In general, I try to resist the urge of rewriting the existing code, for the sake of commit brevity. But, indeed, this is the bug, as the message is printed twice, so I'll fix it.

thaJeztah replied:
Agreed, it's why I mentioned "for a follow up"

kolyshkin commented on daemon/oci_linux.go:
```diff
@@ -316,12 +319,15 @@ func setNamespaces(daemon *Daemon, s *specs.Spec, c *container.Container) error
+	case ipcMode.IsPrivate(), ipcMode.IsShareable():
 		ns := specs.LinuxNamespace{Type: "ipc"}
 		setNamespace(s, ns)
+	default:
```
In general, `""` is valid mode for per-container config, meaning "use the daemon default".

In this context, though, anything that is not `IsContainer()`, `IsHost()`, `IsPrivate()`, or `IsShareable()` is invalid. It can happen for the following reasons:
 - a new mode is added, but is not handled above;
 - somehow a mode string escaped the validation;
 - daemon default is left unset.
If any of these things happen, we have an error here.

kolyshkin replied:
Right, so now I see that yet another reason for `""` appearing here is when a container is created using older version of the daemon, and then started using the new version. We definitely need to handle this case. Will fix.

justincormack commented:
LGTM

I wondered a while back about having a `none` shm mode, as currently it is not possible to have a totally unwriteable container as there is always a `/dev/shm` writeable mount. It is a bit of a niche case (and clearly should never be allowed to be daemon default) but it would be trivial to add now so maybe we should...

thaJeztah commented:
Thanks @kolyshkin changes look good, but https://github.com/moby/moby/pull/34087#discussion_r131108515 has to be addressed before merging, otherwise we break people that are upgrading

kolyshkin commented:
I believe I have addressed all concerns, include the scenario of daemon upgrade and empty IpcMode in container's hostconfig (tested it locally, not sure how to create a test case aside from directly editing hostconfig.json -- let me know if you want me to).

One other thing is, in previous iterations I forgot to add the commit documenting IpcMode in api/swagger.yaml. This one is now included into the main commit, please take a look.

kolyshkin commented:
> I wondered a while back about having a none shm mode, as currently it is not possible to have a totally unwriteable container as there is always a /dev/shm writeable mount

Added a patch, please take a look @justincormack 

I tested it manually, seems working as expected. If we are to go forward with it I'll add a test case.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -665,7 +665,17 @@ definitions:
+                    - `"private"`: own private IPC namespace
+                    - `"shareable"`: own private IPC namespace, with a possibility to share it with other containers
+                    - `"container:<name|id>"`: join another (shareable) container's IPC namespace
+                    - `"host"`: use the host system's IPC namespace
```
For a follow-up PR we can also add an `enum` in the swagger specs; https://github.com/moby/moby/blob/1a0fbc4a6facd35ca93e4eded902e73d0d4b8378/api/swagger.yaml#L171-L174

kolyshkin replied:
I haven't done that as I am not sure how swagger enum can work with `container:<ID>` notation (where `<ID>` is a variable part). I see a few options, but am not sure which one works:
 - omit the `container:<ID>` entrely;
 - specify is as `container:`, omitting the variable suffix;
 - specify is as `container`;
 - specify is as `container:<ID>`.

Glancing through swagger docs (https://swagger.io/docs/specification/enums/) didn't help.

kolyshkin replied:
I mean, is swagger.api is used, for example, to generate something like input parameters validation, omitting `container:` from enum would mean any container:XXX values will be considered invalid and it will break things badly.

thaJeztah replied:
Good point, I didn't consider the `<container>:` case. Another case of bad choices in the past, where UX trickled through in the API; container-ID should've been a separate field

thaJeztah reviewed (commented):
Thanks! Creating a container on current master, then starting the container on a daemon built from this PR now works correctly :+1:

One thing I noticed that has to be fixed is this;

```bash
$ docker run --ipc=none -dit --name bar busybox
docker: --ipc: invalid IPC mode.
```

Looks like there's validation performed in the CLI; [cli/command/container/opts.go#L424-L427]( https://github.com/docker/cli/blob/edfc89f4de7e602bb06729164ae75ab7a90cbef8/cli/command/container/opts.go#L424-L427).

We should remove that, as it's the daemon that should validate if a value is valid (important if the CLI version doesn't match the daemon version)


I just realized there's one thing left to be done; can you add a line to the API version history to mention the new options now availle in the "create" endpont? https://github.com/moby/moby/blob/master/docs/api/version-history.md#v132-api-changes

kolyshkin commented:
> I wondered a while back about having a none shm mode <...>. It is a bit of a niche case

@justincormack I just thought is might not be such a niche case. The fact is most apps are not using /dev/shm, so for most containers this is a useless feature that just adds (not much of, but still) another possible attack vector. When I think about it, maybe it even makes sense to allow "none" to be a daemon default, so the feature have to be explicitly enabled on a per-container basis.

Let me know what do you think.

kolyshkin commented:
> can you add a line to the API version history to mention the new options now availle in the "create" endpont? 

Done (as a separate patch)

thaJeztah reviewed (approved):
LGTM, thanks!!

kolyshkin commented:
> Looks like there's validation performed in the CLI <...> We should remove that

Done (in https://github.com/docker/cli/pull/367)

thaJeztah commented:
ping @justincormack PTAL

kolyshkin commented:
The last push is just a trivial rebase to newer master.

justincormack commented on integration-cli/docker_api_ipcmode_test.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+/* TestAPIIpcModePrivateAndContainer checks that a container created with
+ * --ipc container:ID cat NOT use IPC of another private container.
```
typo cat->can

kolyshkin replied:
One less cat! I feel the urge to compensate! Here:

![kittens-1824367_1920](https://user-images.githubusercontent.com/4522509/29263200-7f35de72-80e0-11e7-91ba-5a4109a796f8.jpg)

kolyshkin commented:
> It could do with a test (make sure it has no mount at least).
Done. Thank you for reviewing!

kolyshkin commented:
Windows test failure seems to be caused by some tests (or execution environment) race, not related to the patchset. Will restart

thaJeztah commented:
All green now; merging 👍
